### DISCUSSION TITLE
tfsec is joining the Trivy family

### DISCUSSION THREAD

**Discussion by itaysk** [aquasecurity/tfsec#1994] [Announcements]
As part of our goal to provide a comprehensive open source security solution for all, we have been consolidating all of our scanning-related efforts in one place, and that is Trivy. Over the past year, tfsec has laid the foundations to Trivy's IaC & misconfigurations scanning capabilities, including Terraform scanning, which has been natively supported in Trivy for a long time now.
Going forward we want to encourage the tfsec community to transition over to Trivy. Moving to Trivy gives you the same excellent Terraform scanning engine, with some extra benefits:
1. Access to more languages and features in the same tool.
2. Access to more integrations with tools and services through the rich ecosystem around Trivy.
3. Commercially supported by Aqua as well as by a the passionate Trivy community.

tfsec will continue to remain available for the time being, although our engineering attention will be directed at Trivy going forward.

**Comment by wilsonmar** [comment]
What are the steps to transition from tfsec to trivy? I did:
brew install trivy
but the command doesn't mention how to scan terraform. I found this video: https://www.youtube.com/watch?v=yf8JC-aNIV8
which recommends this command to scan the current directory with a dot.
trivy conf .
The difference between the two tools are at https://aquasecurity.github.io/trivy/v0.19.2/misconfiguration/comparison/tfsec/
Trivy does not support tfvars and doesn't show statistics counts.

**Reply by simar7** [reply]
looks like you're looking at an older version of the doc. this is the latest release https://aquasecurity.github.io/trivy/v0.37/docs/misconfiguration/scanning/

trivy does support tfvars but does not show statistics counts as tfsec by design.

**Reply by cpflugfelder** [reply]
I noticed that trivy supports tfvars files with the --tf-vars option.  Is there a way to set this in a configuration file?  I'm running into issues making this work with pre-commit.  So far, I have been unable to get trivy to scan using my tfvars file.

**Reply by itaysk** [reply]
@cpflugfelder please ask this question in the Trivy repository so the team can answer

**Comment by AshishDadhich4h2** [comment]
We are using tfsec marketplace task : https://marketplace.visualstudio.com/items?itemName=tfsec.tfsec
Task was updated last 8/2/2022. Is any hope to get dedicated support for this  in near future or it will be deprecated soon?
@simar7 @itaysk

**Reply by itaysk** [reply]
Hi, can you please clarify:
1. is there anything you are missing in the vscode extension? there haven't been much progress since the last release
2. can you consider switching to trivy's extension and doe it address your issue

**Reply by AshishDadhich4h2** [reply]
@itaysk  
Thank you for your quick reply!
We integrated tfsec last week in our ADO pipeline. We are mostly scanning terraform azure based code. We don't have any issue with this so far. 
tfsec extension from market place works great on windows and Linux both self-hosted build agents.
Additionally, it scans private terraform modules that are being called with ssh urls  like git@ssh.dev.azure.com: 

I am happy with tfsec after a quick testing and comparing with trivy's extension.

**Blockers with trivy extension**
1. It doesn't work on windows self-hosted build agent . Error : Unable to locate executable file: 'docker'  
  Means it's expecting docker configuration on windows selfhosted build agent rather downloading windows trivy packages.
2. It doesn`t scan private terraform modules that are being called with ssh urls  like git@ssh.dev.azure.com
3. It doesn't show passed results in ADO trivy UI interface after publishing result and don't find any input option to pass like --include-passed as we use in tfsec.

Are that above known issues?
I tested trivy's extension with my ADO pipeline like
`- task: trivy@1
            displayName: "Compliance Run > trivy"
            inputs:
              path: .
              debug: true
              exitCode: 0`

Please suggest if I missed anything during testing with trivy.

I like following with trivy that are better than tfsec 
1. Better UI interface to show results with categories 
2. It scans secrets that tfsec didn't scan in our case :)



Do you think tfsec is better tool in our case?

Do you think tfsec is better tool in our case?

**Reply by simar7** [reply]
hello @AshishDadhich4h2 - could you confirm what Trivy extension are you using that had those blockers?

**Reply by AshishDadhich4h2** [reply]
@simar7  we are using following ado extension:
https://marketplace.visualstudio.com/items?itemName=AquaSecurityOfficial.trivy-official

**Reply by simar7** [reply]
@giorod3 since you had cut the last release, would you know if the above are supported?

**Comment by shinenelson** [comment]
Is there a migration path for [aquasecurity/tfsec-pr-commenter-action](https://github.com/aquasecurity/tfsec-pr-commenter-action)?

**Reply by steveoh** [reply]
This is the action where I saw the message to come to this discussion. It doesn't appear to have improved much since it was announced a year or two ago now. How is everyone doing?

**Comment by seanturner026** [comment]
Is there any documentation on how `tfsec:ignore` comments should be migrated? 

I found this page but I didn't see anything about comments https://github.com/aquasecurity/tfsec/blob/master/tfsec-to-trivy-migration-guide.md

**Reply by simar7** [reply]
You can find that documented here https://aquasecurity.github.io/trivy/v0.43/docs/configuration/filtering/

**Reply by seanturner026** [reply]
Thanks for that. Is it only possible to filter by the ID? I think filtering by name as is done currently in tfsec has more utility as the ID on its own does not mean anything and requires an additional accompanying comment

https://aquasecurity.github.io/trivy/v0.43/docs/configuration/filtering/#by-inline-comments

**Reply by bodgit** [reply]
Trivy does seem to honour existing tfsec comments, although I agree the names are easier to work with.

**Comment by DBirth** [comment]
Is there more advanced documentation regarding the migration tfsec custom checks to trivy policies (repo policies)?

I tried converting a simple custom check without any success:
```yaml
---
checks:
  - code: missing-default-tags
    description: Custom check to ensure default tags are set on AWS Providers
    requiredTypes:
      - provider
    requiredLabels:
      - aws
    severity: MEDIUM
    matchSpec:
      name: default_tags
      action: isPresent
      subMatch:
        action: and
        predicateMatchSpec:
          - action: contains
            name: tags
            value: TAG_1
          - action: contains
            name: tags
            value: TAG_2
    errorMessage: |
      The missing tags helps structure .....
```

From what I see in the code when initialising the repo scanner, the terraform always sets the source to cloud (rego is initialized with source rego):
https://github.com/aquasecurity/trivy-iac/blob/6d4668a0a3cf9d959c086fccffb3112da06e0015/pkg/scanners/terraform/scanner.go#L151-L152
```go 
regoScanner := rego.NewScanner(types.SourceCloud, s.options...)
regoScanner.SetParentDebugLogger(s.debug)
```
Also before applying the rego rules, an adapt is applied:
https://github.com/aquasecurity/trivy-iac/blob/6d4668a0a3cf9d959c086fccffb3112da06e0015/pkg/scanners/terraform/executor/executor.go#L88C1-L91
```go
adaptationTime := time.Now()
infra := adapter.Adapt(modules)
metrics.Timings.Adaptation = time.Since(adaptationTime)
e.debug.Log("Adapted %d module(s) into defsec state data.", len(modules))
```

And at this point seems that we're bounded to what we can get, and it doesn't seem possible to actually validate default tags.
Am I missing some sort of configuration?

**Reply by itaysk** [reply]
@simar7 wdyt?

**Reply by simar7** [reply]
> I tried converting a simple custom check without any success:

@DBirth - can you share your converted rego policy?

**Reply by DBirth** [reply]
Hi @simar7 ,
I didn't complete because I couldn't find how to check the provider default tags.
But I got to here:
```rego
# METADATA
# title: "Mandatory default tags"
# description: "Tags are mandatory for all projects"
# scope: package
# custom:
#   id: CUSTOM-AWS-001
#   severity: LOW
#   short_code: missing_mandatory_tags
#   recommended_action: "Ensure that tags [TAG_1, TAG_2] exist"
package user.terraform. missing_mandatory_tags

deny[res] {
    trace(sprintf("Check input object! %v", [input]))
    1 == 1
    msg := "Checking rego input to implement next steps"
    res := result.new(msg, input.aws.config)
}
```

But in the json returned by the trace I couldn't see a way to get the provider.

**Reply by simar7** [reply]
Sorry but could you help me understand what are you trying to check? In Trivy we have the concept of providers (like AWS) that have services (like EC2, RDS, etc.) which further have attributes to them that can be scanned for. 

For example take a look at the definition of an RDS instance https://github.com/aquasecurity/defsec/blob/v0.93.1/pkg/providers/aws/rds/rds.go based on this we can write such a rego policy https://github.com/aquasecurity/defsec/blob/v0.93.1/rules/cloud/policies/aws/rds/disable_cluster_skip_final_snapshot.rego

**Reply by DBirth** [reply]
So basically I want to validate the arguments at the provider level, for example:
```hcl
provider "aws" {
   region = "eu-west-1"
   default_tags {
      tags = {
         "TAG": "VALUE"
      }
   }
}
```
Here I would like to set a custom check that enforces that the only allowed region is `eu-west-1`, and that it should have the following default tags `TAG_1` and `TAG_2`.

As you stated, trivy has the concept of providers and for AWS the model (input) is https://github.com/aquasecurity/defsec/blob/v0.93.1/pkg/providers/aws/aws.go.

But there's nothing that allows me to validate the provider block from what I can see, trivy only has mappings for the individual resources. But the general ones like for `region`, and `default_tags` seems to be missing.

**Reply by simar7** [reply]
@DBirth you're right we don't have support for that today. I've cut https://github.com/aquasecurity/trivy/issues/5459 to track adding this.

**Comment by nicolamarangoni** [comment]
Hello, what about trivy with pre-commit as replacement of tfsec?
Here: https://github.com/mxab/pre-commit-trivy
I find only hooks for docker, should we wait until we have hooks for terraform?

**Reply by simar7** [reply]
It is not an Aqua open source maintained project so we can't comment on it.

**Comment by antigenius0910** [comment]
@itaysk 

Here is an open issue and open PR for approval if Aqua want tfsec people start moving to Trivy. thank you!   
https://github.com/aquasecurity/trivy-action/pull/311
